File: agentpit/datastructures/market.py
from typing import Optional, Any
from pydantic import BaseModel, field_validator

from agentpit.common import check_state
from agentpit.datastructures.condition_id import ConditionId
from agentpit.db.table_create import MarketState
from agentpit.polymarket.conditional_token_framework import ConditionalTokenFramework
from agentpit.utils.parse import is_hex256


class Market(BaseModel):
    question: str
    slug: str
    market_id: int
    polymarket_id: Optional[int] = None
    polymarket_condition_id: Optional[str] = None
    condition_id: ConditionId = None
    description: str
    erc1155_tokens: list[tuple[str, str]]
    start_date: int
    end_date: Optional[int | None]
    resolved_outcome: Optional[int] = None
    market_state: MarketState
    event_id: Optional[int] = None
    outcome_label: Optional[str] = None
    icon_url: Optional[str] = None

    def model_post_init(self, __context):
        check_state(self.market_state != MarketState.RESOLVED or self.resolved_outcome is not None,
                    "Resolved market must have an outcome")
        check_state(len(self.question) > 0,
                    "Question must not be empty")
        check_state(len(self.erc1155_tokens) > 0,
                    "Must have at least one ERC1155 token")
        check_state(self.end_date is None or self.end_date > self.start_date,
                    "End date must be after start date")
        check_state(self.condition_id is not None)

File: ui/src/types/market.ts
export type MarketState =
  | "DRAFT"
  | "ACTIVE"
  | "CLOSED"
  | "RESOLVED"
  | "CANCELLED";

export type Erc1155Token = readonly [tokenId: string, label: string];

export interface Market {
  market_id: number;
  question: string;
  slug: string;
  description: string;
  erc1155_tokens: Erc1155Token[];
  start_date: number | null;
  end_date: number | null;
  market_state: MarketState;
  resolved_outcome: number | null;
  polymarket_id: number | null;
  condition_id: string;
  event_id: number | null;
  outcome_label: string | null;
  icon_url: string | null;
  /** One probability in [0, 1] per outcome, index-aligned with
   *  `erc1155_tokens`. Empty when the payload carried no usable prices. */
  outcome_prices: number[];
  /** Best bid / ask for outcome[0] (YES) in [0, 1] — the server ships exactly
   *  one pair per market (Gamma parity). `null` means no resting order on that
   *  side; the wire sends 0.0 for that case and it must not reach the UI as a
   *  real 0¢ quote. */
  best_bid: number | null;
  best_ask: number | null;
}

export interface ListMarketsResponse {
  markets: Market[];
  total: number;
  limit: number;
  offset: number;
}

export interface SparklinePoint {
  /** Trade timestamp in unix seconds. */
  t: number;
  /** Probability in [0, 1]. */
  p: number;
}

export interface PricesHistoryResponse {
  history: SparklinePoint[];
}
